File: sdks/go/signature.go
// Package webhookrelay provides a receiver-side HMAC verifier for the
// Webhook Relay API's X-Webhook-Signature header.
//
// Header format: t={unix_ts},v1={hex_hmac_sha256}
// Signed payload: "{t}.{raw_body}"
//
// The raw body must be the exact bytes received. json.Marshal(json.Unmarshal(...))
// is NOT safe — Go's encoding/json will reorder map keys and break the signature.
package webhookrelay

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"strconv"
	"strings"
	"time"
)

// VerifySignature returns true if the header signature matches body+secret
// within tolerance. Default tolerance is 5 minutes; pass 0 to use it.
func VerifySignature(secret, body, header string, tolerance time.Duration) bool {
	if header == "" {
		return false
	}
	if tolerance == 0 {
		tolerance = 5 * time.Minute
	}

	var ts int64
	var sig string
	for _, segment := range strings.Split(header, ",") {
		kv := strings.SplitN(strings.TrimSpace(segment), "=", 2)
		if len(kv) != 2 {
			continue
		}
		switch kv[0] {
		case "t":
			parsed, err := strconv.ParseInt(kv[1], 10, 64)
			if err != nil {
				return false
			}
			ts = parsed
		case "v1":
			sig = kv[1]
		}
	}
	if ts == 0 || sig == "" {
		return false
	}

	delta := time.Since(time.Unix(ts, 0))
	if delta < 0 {
		delta = -delta
	}
	if delta > tolerance {
		return false
	}

	mac := hmac.New(sha256.New, []byte(secret))
	fmt.Fprintf(mac, "%d.%s", ts, body)
	expected := hex.EncodeToString(mac.Sum(nil))

	return hmac.Equal([]byte(expected), []byte(sig))
}

File: sdks/python/webhook_relay/signature.py
"""Receiver-side HMAC verifier for X-Webhook-Signature.

Header format: ``t={unix_ts},v1={hex_hmac_sha256}``
Signed payload: ``"{t}.{raw_body}"``

The raw body must be the exact bytes received. ``json.dumps(json.loads(...))``
is NOT safe — Python's serializer may reorder keys or change whitespace,
breaking the signature.
"""

from __future__ import annotations

import hashlib
import hmac
import time
from typing import Optional, Union


def verify_signature(
    secret: str,
    body: Union[str, bytes],
    header: Optional[str],
    tolerance_seconds: int = 300,
) -> bool:
    """Return True if the header signature matches body+secret within tolerance."""
    if not header:
        return False

    parts: dict[str, str] = {}
    for segment in header.split(","):
        kv = segment.strip().split("=", 1)
        if len(kv) == 2:
            parts[kv[0]] = kv[1]

    if "t" not in parts or "v1" not in parts:
        return False

    try:
        ts = int(parts["t"])
    except ValueError:
        return False

    if abs(int(time.time()) - ts) > tolerance_seconds:
        return False

    body_bytes = body.encode("utf-8") if isinstance(body, str) else body
    signed = f"{ts}.".encode("utf-8") + body_bytes
    expected = hmac.new(secret.encode("utf-8"), signed, hashlib.sha256).hexdigest()

    return hmac.compare_digest(expected, parts["v1"])
